internal/wal/wal_aof.go

```diff
@@ -98,85 +384,90 @@ func checksum(command string) []byte {
 	return hash[:]
 }
 
+// ForEachCommand reads all the logged commands from WAL segments and applies a given function to each command.
 func (w *WALAOF) ForEachCommand(f func(c cmd.DiceDBCmd) error) error {
-	var length uint32
-
-	files, err := os.ReadDir(w.logDir)
+	files, err := filepath.Glob(filepath.Join(w.logDir, segmentPrefix+"*"))
 	if err != nil {
 		return fmt.Errorf("failed to read log directory: %v", err)
 	}
 
-	var walFiles []os.DirEntry
-
-	for _, file := range files {
-		if !file.IsDir() && filepath.Ext(file.Name()) == ".aof" {
-			walFiles = append(walFiles, file)
-		}
-	}
-
-	if len(walFiles) == 0 {
+	if len(files) == 0 {
 		return fmt.Errorf("no valid WAL files found in log directory")
 	}
 
-	// Sort files by timestamp in ascending order
-	sort.Slice(walFiles, func(i, j int) bool {
-		timestampStrI := walFiles[i].Name()[4:17]
-		timestampStrJ := walFiles[j].Name()[4:17]
-		timestampI, errI := time.Parse("20060102_1504", timestampStrI)
-		timestampJ, errJ := time.Parse("20060102_1504", timestampStrJ)
-		if errI != nil || errJ != nil {
-			return false
-		}
-		return timestampI.Before(timestampJ)
+	sort.Slice(files, func(i, j int) bool {
+		idxI := getSegmentIndex(files[i])
+		idxJ := getSegmentIndex(files[j])
+		return idxI < idxJ
 	})
 
-	for _, file := range walFiles {
-		filePath := filepath.Join(w.logDir, file.Name())
-
-		slog.Debug("loading WAL", slog.Any("file", filePath))
+	readBuf := bufferPool.Get().([]byte)
+	defer bufferPool.Put(readBuf)
 
-		file, err := os.OpenFile(filePath, os.O_RDONLY, 0644)
+	for _, file := range files {
+		fHandle, err := os.OpenFile(file, os.O_RDONLY, 0644)
 		if err != nil {
-			return fmt.Errorf("failed to open WAL file %s: %v", file.Name(), err)
+			return fmt.Errorf("failed to open WAL file %s: %v", file, err)
 		}
 
+		reader := bufio.NewReader(fHandle)
+
 		for {
-			if err := binary.Read(file, binary.BigEndian, &length); err != nil {
+			lengthBuf := make([]byte, 4)
+			if _, err := io.ReadFull(reader, lengthBuf); err != nil {
 				if err == io.EOF {
 					break
 				}
 				return fmt.Errorf("failed to read entry length: %v", err)
 			}
+			length := binary.BigEndian.Uint32(lengthBuf)
 
-			// TODO: Optimize this allocation.
-			// Pre-allocate and reuse rather than allocating for each entry.
-			readBufBytes := make([]byte, length)
-			if _, err := io.ReadFull(file, readBufBytes); err != nil {
+			if cap(readBuf) < int(length) {
+				readBuf = make([]byte, length)
+			} else {
+				readBuf = readBuf[:length]
+			}
+
+			if _, err := io.ReadFull(reader, readBuf); err != nil {
 				return fmt.Errorf("failed to read entry data: %v", err)
 			}
 
 			entry := &WALLogEntry{}
-			if err := proto.Unmarshal(readBufBytes, entry); err != nil {
+			if err := proto.Unmarshal(readBuf, entry); err != nil {
```

Comment:
This should probably be a design discussion, but not sure if we need the multi language and platform compatibility by using proto here when we can write the bytes directly to the file. This would make the WAL faster and another API can be built on top of the WAL for subscribers who want to read the WAL.

Reply:
cc @JyotinderSingh @soumya-codes